backup-s3/src/main/scala/io/aiven/guardian/kafka/backup/s3/BackupClient.scala

```diff
@@ -192,24 +201,43 @@ class BackupClient[T <: KafkaClientInterface](maybeS3Settings: Option[S3Settings
   override def backupToStorageTerminateSink(
       previousState: PreviousState
   ): Sink[ByteString, Future[BackupResult]] = {
-    logger.info(
-      s"Terminating and completing previous backup with key: ${previousState.previousKey} and uploadId:${previousState.state.uploadId}"
-    )
-    val sink = S3
-      .resumeMultipartUploadWithHeaders(
-        s3Config.dataBucket,
-        previousState.previousKey,
-        previousState.state.uploadId,
-        previousState.state.parts,
-        s3Headers = s3Headers,
-        chunkingParallelism = 1
-      )
+    implicit val ec: ExecutionContext = system.dispatcher
 
-    val base = sink.mapMaterializedValue(future => future.map(result => Some(result))(ExecutionContext.parasitic))
+    RestartSink.withBackoff(
+      s3Config.errorRestartSettings
+    ) { () =>
+      Sink.lazyFutureSink { () =>
+        for {
+          exists <- checkObjectExists(previousState.previousKey)
+        } yield
+          if (exists) {
```

Comment:
Hm ... shouldn't the condition be reversed here? `if (!exists)` ?

Reply:
Nope this is the correct condition, to clarify this piece of code rectifying one of the issues with eventual consistency for multipart uploads on S3.

The reason why we are checking if an object exists in this spot is that if it an S3 object does exist we shouldn't even be in this specific code path (hence the `Sink.ignore`), basically whats happening is that we make an earlier call which lists all multipart uploads and in some cases that multipart upload has already been completed and is actually now an object. Ideally in if an S3 object exists there shouldn't be an in progress multipart upload but whats the eventual consistency problem is.

Reply:
Oh I see, so `exists` is about S3 object whereas we looking for uploads, got it

Reply:
Yes precisely